lib/que/locker.rb

```diff
@@ -24,9 +24,23 @@ def initialize(queue:, cursor_expiry:, metrics:)
     def with_locked_job
       reset_cursor if cursor_expired?
 
+      job = lock_job
+
+      # Becuase we were using a cursor when we tried to lock this job, if we fail to find
+      # a job it is not necessarily the case that there aren't jobs in the queue. We may
+      # have been excluding candidate jobs due to the cursor that are now due to be
+      # worked.
+      #
+      # We should attempt to lock again after resetting our cursor to make sure we include
+      # jobs that were excluded in the first attempt.
+      if job.nil? && @cursor
+        reset_cursor
+        job = lock_job
+      end
+
       # Check that the job hasn't just been worked by another worker (it's possible to
       # lock a job that's just been destroyed because pg locks don't obey MVCC). If it has
-      # been worked, act as if we've worked it.
+      # been worked, return nil to immediately retry locking.
```

Comment:
I think it'd be useful to clarify in a comment at the top of this function that "returning nil" will get this method called again immediately, while "yielding nil" makes the worker sleep a bit and then call this method again.

The distinction is a bit subtle, and this particular idea commented here is not that obvious (at least to me, I had to go look at `worker.rb` to make sure I was understanding this correctly)